### ISSUE TITLE
FindOpenCV.cmake seems to be missing

### ISSUE THREAD

**Issue by linas** [hansonrobotics/CppMT#6]
When building, I get this error:

```
CMake Error at CMakeLists.txt:9 (find_package):
  By not providing "FindOpenCV.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "OpenCV", but
  CMake did not find one.

  Could not find a package configuration file provided by "OpenCV" with any
  of the following names:

    OpenCVConfig.cmake
    opencv-config.cmake

  Add the installation prefix of "OpenCV" to CMAKE_PREFIX_PATH or set
  "OpenCV_DIR" to a directory containing one of the above files.  If "OpenCV"
  provides a separate development package or SDK, be sure it has been
  installed.


-- Configuring incomplete, errors occurred!
See also "/opt/hansonrobotics/vision/CppMT/CMakeFiles/CMakeOutput.log".
```

**Comment by linas** [comment]
There seems to be one here: http://code.opencv.org/attachments/196/FindOpenCV.cmake

**Comment by linas** [comment]
Copying the file at the above URL into the repo, and making the change below seems enough to solve this bug.

```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index a94b708..99cd88e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -6,6 +6,8 @@ option(BUILD_SHARED_LIBS "Build shared libraries." ON)
 option(BUILD_WRAPPER "Build C and Python wrapper." ON)
 option(BUILD_TRAX_CLIENT "Build the trax client." OFF)

+SET(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}")
+
 find_package(OpenCV REQUIRED)
 find_package(Boost REQUIRED COMPONENTS system thread filesystem)
 # uninstall target
```

**Comment by linas** [comment]
Actually, above is not enough. Running that version of FindOpenCV results in errors having to do with extracting the version numbers.  Stubbing out lines 120,121 and 122 results in the error

```
CMake Error at FindOpenCV.cmake:267 (message):
  OpenCV required but some headers or libs not found.
Call Stack (most recent call first):
  CMakeLists.txt:11 (find_package)


-- Configuring incomplete, errors occurred!
```

stubbing out that error, and continuing onwards results in a failed build:

```
[100%] Linking CXX executable cmt
CMakeFiles/cmt.dir/gui.cpp.o: In function `screenLog(cv::Mat, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)':
gui.cpp:(.text+0x38): undefined reference to `cv::getTextSize(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, double, int, int*)'
gui.cpp:(.text+0xac): undefined reference to `cv::rectangle(cv::Mat&, cv::Point_<int>, cv::Point_<int>, cv::Scalar_<double> const&, int, int, int)'
gui.cpp:(.text+0x107): undefined reference to `cv::putText(cv::Mat&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, cv::Point_<int>, int, double, cv::Scalar_<double>, int, int, bool)'
CMakeFiles/cmt.dir/gui.cpp.o: In function `onMouse(int, int, int, int, void*)':
gui.cpp:(.text+0x1e6): undefined reference to `cv::_OutputArray::_OutputArray(cv::Mat&)'
gui.cpp:(.text+0x1f7): undefined reference to `cv::Mat::copyTo(cv::_OutputArray const&) const'
gui.cpp:(.text+0x280): undefined reference to `cv::rectangle(cv::Mat&, cv::Point_<int>, cv::Point_<int>, cv::Scalar_<double> const&, int, int, int)'
gui.cpp:(.text+0x2a2): undefined reference to `cv::_InputArray::_InputArray(cv::Mat const&)'
gui.cpp:(.text+0x2b3): undefined reference to `cv::imshow(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, cv::_InputArray const&)'
gui.cpp:(.text+0x2d1): undefined reference to `cv::Mat::deallocate()'
gui.cpp:(.text+0x331): undefined reference to `cv::fastFree(void*)'
gui.cpp:(.text+0x579): undefined reference to `cv::fastFree(void*)'
gui.cpp:(.text+0x7a5): undefined reference to `cv::fastFree(void*)'
gui.cpp:(.text+0x9aa): undefined reference to `cv::fastFree(void*)'
gui.cpp:(.text+0x9e9): undefined reference to `cv::Mat::copySize(cv::Mat const&)'
gui.cpp:(.text+0xa0d): undefined reference to `cv::Mat::copySize(cv::Mat const&)'
gui.cpp:(.text+0xa28): undefined reference to `cv::Mat::deallocate()'
gui.cpp:(.text+0xa3c): undefined reference to `cv::Mat::deallocate()'
gui.cpp:(.text+0xa61): undefined reference to `cv::Mat::copySize(cv::Mat const&)'
gui.cpp:(.text+0xa78): undefined reference to `cv::Mat::deallocate()'
CMakeFiles/cmt.dir/gui.cpp.o: In function `getRect(cv::Mat, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)':
gui.cpp:(.text+0xbe5): undefined reference to `cv::Mat::copySize(cv::Mat const&)'
gui.cpp:(.text+0xc26): undefined reference to `cv::setMouseCallback(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, void (*)(int, int, int, int, void*), void*)'
gui.cpp:(.text+0xc4e): undefined reference to `cvWaitKey'
gui.cpp:(.text+0xc63): undefined reference to `cv::setMouseCallback(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, void (*)(int, int, int, int, void*), void*)'
gui.cpp:(.text+0xc7f): undefined reference to `cv::Mat::deallocate()'
gui.cpp:(.text+0xd56): undefined reference to `cv::Mat::deallocate()'
CMakeFiles/cmt.dir/gui.cpp.o: In function `cv::Mat::~Mat()':
gui.cpp:(.text._ZN2cv3MatD2Ev[_ZN2cv3MatD5Ev]+0x14): undefined reference to `cv::Mat::deallocate()'
gui.cpp:(.text._ZN2cv3MatD2Ev[_ZN2cv3MatD5Ev]+0x59): undefined reference to `cv::fastFree(void*)'
CMakeFiles/cmt.dir/main.cpp.o: In function `display(cv::Mat, cmt::CMT&)':
main.cpp:(.text+0x2a3): undefined reference to `cv::circle(cv::Mat&, cv::Point_<int>, int, cv::Scalar_<double> const&, int, int, int)'
main.cpp:(.text+0x2ea): undefined reference to `cv::RotatedRect::points(cv::Point_<float>*) const'
main.cpp:(.text+0x379): undefined reference to `cv::line(cv::Mat&, cv::Point_<int>, cv::Point_<int>, cv::Scalar_<double> const&, int, int, int)'
main.cpp:(.text+0x390): undefined reference to `cv::_InputArray::_InputArray(cv::Mat const&)'
main.cpp:(.text+0x39d): undefined reference to `cv::imshow(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, cv::_InputArray const&)'
main.cpp:(.text+0x3a7): undefined reference to `cv::waitKey(int)'
CMakeFiles/cmt.dir/main.cpp.o: In function `write_rotated_rect[abi:cxx11](cv::RotatedRect)':
main.cpp:(.text+0x421): undefined reference to `cv::RotatedRect::points(cv::Point_<float>*) const'
CMakeFiles/cmt.dir/main.cpp.o: In function `cv::BFMatcher::~BFMatcher()':
main.cpp:(.text._ZN2cv9BFMatcherD0Ev[_ZN2cv9BFMatcherD5Ev]+0x7): undefined reference to `vtable for cv::BFMatcher'
main.cpp:(.text._ZN2cv9BFMatcherD0Ev[_ZN2cv9BFMatcherD5Ev]+0xc): undefined reference to `cv::DescriptorMatcher::~DescriptorMatcher()'
CMakeFiles/cmt.dir/main.cpp.o: In function `cv::Mat::Mat(cv::Mat const&)':
main.cpp:(.text._ZN2cv3MatC2ERKS0_[_ZN2cv3MatC5ERKS0_]+0x98): undefined reference to `cv::Mat::copySize(cv::Mat const&)'
CMakeFiles/cmt.dir/main.cpp.o: In function `cv::Mat::operator=(cv::Mat const&)':
main.cpp:(.text._ZN2cv3MataSERKS0_[_ZN2cv3MataSERKS0_]+0x81): undefined reference to `cv::Mat::copySize(cv::Mat const&)'
main.cpp:(.text._ZN2cv3MataSERKS0_[_ZN2cv3MataSERKS0_]+0xf4): undefined reference to `cv::Mat::deallocate()'
CMakeFiles/cmt.dir/main.cpp.o: In function `cv::BFMatcher::~BFMatcher()':
main.cpp:(.text._ZN2cv9BFMatcherD2Ev[_ZN2cv9BFMatcherD5Ev]+0x3): undefined reference to `vtable for cv::BFMatcher'
main.cpp:(.text._ZN2cv9BFMatcherD2Ev[_ZN2cv9BFMatcherD5Ev]+0x8): undefined reference to `cv::DescriptorMatcher::~DescriptorMatcher()'
CMakeFiles/cmt.dir/main.cpp.o: In function `cmt::Matcher::~Matcher()':
main.cpp:(.text._ZN3cmt7MatcherD2Ev[_ZN3cmt7MatcherD5Ev]+0x7c): undefined reference to `cv::fastFree(void*)'
main.cpp:(.text._ZN3cmt7MatcherD2Ev[_ZN3cmt7MatcherD5Ev]+0xb8): undefined reference to `vtable for cv::BFMatcher'
main.cpp:(.text._ZN3cmt7MatcherD2Ev[_ZN3cmt7MatcherD5Ev]+0xc0): undefined reference to `cv::DescriptorMatcher::~DescriptorMatcher()'
main.cpp:(.text._ZN3cmt7MatcherD2Ev[_ZN3cmt7MatcherD5Ev]+0xd4): undefined reference to `cv::fastFree(void*)'
main.cpp:(.text._ZN3cmt7MatcherD2Ev[_ZN3cmt7MatcherD5Ev]+0xe5): undefined reference to `cv::Mat::deallocate()'
CMakeFiles/cmt.dir/main.cpp.o: In function `cmt::Consensus::~Consensus()':
main.cpp:(.text._ZN3cmt9ConsensusD2Ev[_ZN3cmt9ConsensusD5Ev]+0x77): undefined reference to `cv::fastFree(void*)'
main.cpp:(.text._ZN3cmt9ConsensusD2Ev[_ZN3cmt9ConsensusD5Ev]+0xca): undefined reference to `cv::fastFree(void*)'
main.cpp:(.text._ZN3cmt9ConsensusD2Ev[_ZN3cmt9ConsensusD5Ev]+0xf0): undefined reference to `cv::Mat::deallocate()'
main.cpp:(.text._ZN3cmt9ConsensusD2Ev[_ZN3cmt9ConsensusD5Ev]+0x105): undefined reference to `cv::Mat::deallocate()'
CMakeFiles/cmt.dir/main.cpp.o: In function `cmt::CMT::~CMT()':
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x77): undefined reference to `cv::fastFree(void*)'
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x160): undefined reference to `cv::fastFree(void*)'
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x22b): undefined reference to `cv::fastFree(void*)'
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x29d): undefined reference to `cv::fastFree(void*)'
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x32c): undefined reference to `cv::fastFree(void*)'
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x374): undefined reference to `vtable for cv::BFMatcher'
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x37c): undefined reference to `cv::DescriptorMatcher::~DescriptorMatcher()'
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x390): undefined reference to `cv::fastFree(void*)'
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x3ba): undefined reference to `cv::fastFree(void*)'
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x3e2): undefined reference to `cv::fastFree(void*)'
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x3f5): undefined reference to `cv::Mat::deallocate()'
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x408): undefined reference to `cv::Mat::deallocate()'
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x420): undefined reference to `cv::Mat::deallocate()'
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x438): undefined reference to `cv::Mat::deallocate()'
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x450): undefined reference to `cv::Mat::deallocate()'
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x611): undefined reference to `cv::fastFree(void*)'
main.cpp:(.text._ZN3cmt3CMTD2Ev[_ZN3cmt3CMTD5Ev]+0x634): undefined reference to `cv::fastFree(void*)'
CMakeFiles/cmt.dir/main.cpp.o: In function `main':
main.cpp:(.text.startup+0xc4a): undefined reference to `cv::namedWindow(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)'
main.cpp:(.text.startup+0xc57): undefined reference to `cv::VideoCapture::VideoCapture()'
main.cpp:(.text.startup+0xc75): undefined reference to `cv::VideoCapture::open(int)'
main.cpp:(.text.startup+0xc87): undefined reference to `cv::VideoCapture::isOpened() const'
main.cpp:(.text.startup+0xd45): undefined reference to `cv::VideoCapture::operator>>(cv::Mat&)'
main.cpp:(.text.startup+0xdab): undefined reference to `cv::_InputArray::_InputArray(cv::Mat const&)'
main.cpp:(.text.startup+0xdbd): undefined reference to `cv::imshow(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, cv::_InputArray const&)'
main.cpp:(.text.startup+0xdc7): undefined reference to `cv::waitKey(int)'
main.cpp:(.text.startup+0xe70): undefined reference to `cv::VideoCapture::operator>>(cv::Mat&)'
main.cpp:(.text.startup+0x1046): undefined reference to `cv::_OutputArray::_OutputArray(cv::Mat&)'
main.cpp:(.text.startup+0x105e): undefined reference to `cv::_InputArray::_InputArray(cv::Mat const&)'
main.cpp:(.text.startup+0x1070): undefined reference to `cv::cvtColor(cv::_InputArray const&, cv::_OutputArray const&, int, int)'
main.cpp:(.text.startup+0x1232): undefined reference to `cv::_OutputArray::_OutputArray(cv::Mat&)'
main.cpp:(.text.startup+0x1242): undefined reference to `cv::Mat::copyTo(cv::_OutputArray const&) const'
main.cpp:(.text.startup+0x12db): undefined reference to `cv::VideoCapture::~VideoCapture()'
main.cpp:(.text.startup+0x153b): undefined reference to `cv::_OutputArray::_OutputArray(cv::Mat&)'
main.cpp:(.text.startup+0x1550): undefined reference to `cv::_InputArray::_InputArray(cv::Mat const&)'
main.cpp:(.text.startup+0x1567): undefined reference to `cv::cvtColor(cv::_InputArray const&, cv::_OutputArray const&, int, int)'
main.cpp:(.text.startup+0x170f): undefined reference to `cv::VideoCapture::operator>>(cv::Mat&)'
main.cpp:(.text.startup+0x1932): undefined reference to `cv::VideoCapture::get(int)'
main.cpp:(.text.startup+0x1bee): undefined reference to `cv::VideoCapture::open(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
main.cpp:(.text.startup+0x1c10): undefined reference to `cv::VideoCapture::set(int, double)'
main.cpp:(.text.startup+0x1c2f): undefined reference to `cv::VideoCapture::set(int, double)'
main.cpp:(.text.startup+0x1c41): undefined reference to `cv::VideoCapture::get(int)'
main.cpp:(.text.startup+0x1f08): undefined reference to `cv::imread(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)'
main.cpp:(.text.startup+0x1fb7): undefined reference to `cv::_OutputArray::_OutputArray(cv::Mat&)'
main.cpp:(.text.startup+0x1fcc): undefined reference to `cv::_InputArray::_InputArray(cv::Mat const&)'
main.cpp:(.text.startup+0x1fe0): undefined reference to `cv::cvtColor(cv::_InputArray const&, cv::_OutputArray const&, int, int)'
main.cpp:(.text.startup+0x2293): undefined reference to `cv::imread(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)'
main.cpp:(.text.startup+0x2326): undefined reference to `cv::_OutputArray::_OutputArray(cv::Mat&)'
main.cpp:(.text.startup+0x2336): undefined reference to `cv::_InputArray::_InputArray(cv::Mat const&)'
main.cpp:(.text.startup+0x2348): undefined reference to `cv::cvtColor(cv::_InputArray const&, cv::_OutputArray const&, int, int)'
main.cpp:(.text.startup+0x23d3): undefined reference to `cv::RotatedRect::boundingRect() const'
main.cpp:(.text.startup+0x261c): undefined reference to `cv::VideoCapture::get(int)'
main.cpp:(.text.startup+0x27c3): undefined reference to `cv::VideoCapture::~VideoCapture()'
libcmt.so: undefined reference to `cv::DescriptorExtractor::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
libcmt.so: undefined reference to `cv::FeatureDetector::detect(cv::Mat const&, std::vector<cv::KeyPoint, std::allocator<cv::KeyPoint> >&, cv::Mat const&) const'
libcmt.so: undefined reference to `vtable for cv::_OutputArray'
libcmt.so: undefined reference to `cv::Mat::create(int, int const*, int)'
libcmt.so: undefined reference to `cv::DescriptorMatcher::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
libcmt.so: undefined reference to `vtable for cv::_InputArray'
libcmt.so: undefined reference to `cv::DescriptorExtractor::compute(cv::Mat const&, std::vector<cv::KeyPoint, std::allocator<cv::KeyPoint> >&, cv::Mat&) const'
libcmt.so: undefined reference to `cv::DescriptorMatcher::knnMatch(cv::Mat const&, cv::Mat const&, std::vector<std::vector<cv::DMatch, std::allocator<cv::DMatch> >, std::allocator<std::vector<cv::DMatch, std::allocator<cv::DMatch> > > >&, int, cv::Mat const&, bool) const'
libcmt.so: undefined reference to `cv::Mat::Mat(cv::Mat const&, cv::Rect_<int> const&)'
libcmt.so: undefined reference to `cv::_OutputArray::_OutputArray(cv::Mat const&)'
libcmt.so: undefined reference to `cv::FeatureDetector::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
libcmt.so: undefined reference to `cv::calcOpticalFlowPyrLK(cv::_InputArray const&, cv::_InputArray const&, cv::_InputArray const&, cv::_OutputArray const&, cv::_OutputArray const&, cv::_OutputArray const&, cv::Size_<int>, int, cv::TermCriteria, int, double)'
libcmt.so: undefined reference to `cv::vconcat(cv::_InputArray const&, cv::_InputArray const&, cv::_OutputArray const&)'
libcmt.so: undefined reference to `cv::Mat::Mat(cv::Mat const&, cv::Range const&, cv::Range const&)'
collect2: error: ld returned 1 exit status
CMakeFiles/cmt.dir/build.make:128: recipe for target 'cmt' failed
make[2]: *** [cmt] Error 1
CMakeFiles/Makefile2:173: recipe for target 'CMakeFiles/cmt.dir/all' failed
make[1]: *** [CMakeFiles/cmt.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by linas** [comment]
I am guessing that the above problem occurs, because `${OpenCV_LIBS}` is never set, and so lines such as this:

```
target_link_libraries(cmtlib ${OpenCV_LIBS} ${Boost_LIBRARIES})
```

end up failing.  So, basically, that version of `FindOpenCV.cmake` is bad.  Maybe there is another version that works?

**Comment by linas** [comment]
This version:
https://github.com/mikeroberts3000/EfficientHierarchicalGraphBasedVideoSegmentationExporter/blob/master/code/segmentation_exporter/cmake/FindOpenCV.cmake

when running `cmake .`, it works without errors. However, make fails in the final link stage, as before.

**Comment by wenwei-dev** [comment]
How about install libopencv-dev and run again? It could be the opencv is not installed to the desired directory, so cmake can't find it.

**Comment by linas** [comment]
I'm stumped. See https://bugs.launchpad.net/ubuntu/+source/opencv/+bug/1167014

**Comment by linas** [comment]
I'm utterly confused. After spending the last 3 hours trying to debug this, and consistently failing  ... repeatedly, ... for hours ... it now just ... started working.  

What made is start working is that I ran `cmake --trace .`  once, and after doing this, it just started working.  I had been doing clean builds, by removing `CMakeCache.txt` and `CMakeFiles`, etc. but always got the above bugs. But after running `cmake --trace .`  just once, the problem goes away .. completely, and becomes unreproducible.  

I really don't get it. Closing.
